### PR TITLE
wip - Add guessed typed information to flyteremote objects

### PR DESCRIPTION
**Pull request by eapolinario**
Signed-off-by: Eduardo Apolinario <curupa@gmail.com>

# TL;DR
Tasks and workflows loaded via remote do not contain enough typing information to allow for the reconstruction of the original types. This PR augments the `flyteremote` types with typing information as defined by the user.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -407,17 +407,18 @@ def literal_map_to_kwargs(
         return {k: TypeEngine.to_python_value(ctx, lm.literals[k], v) for k, v in python_types.items()}
 
     @classmethod
-    def dict_to_literal_map(cls, ctx: FlyteContext, d: typing.Dict[str, typing.Any]) -> LiteralMap:
+    def dict_to_literal_map(cls, ctx: FlyteContext, d: typing.Dict[str, typing.Any], m: typing.Dict[str, type]) -> LiteralMap:
```
Comment (wild-endeavor) [inline flytekit/core/type_engine.py:410]: Let's make the new argument optional - this is presumably a public API (though I doubt very many people use it), so we should keep the api backwards compatible.

File: flytekit/core/type_engine.py [hunk 2]
```diff
@@ -407,17 +407,26 @@ def literal_map_to_kwargs(
         return {k: TypeEngine.to_python_value(ctx, lm.literals[k], v) for k, v in python_types.items()}
 
     @classmethod
-    def dict_to_literal_map(cls, ctx: FlyteContext, d: typing.Dict[str, typing.Any]) -> LiteralMap:
+    def dict_to_literal_map(
+        cls, ctx: FlyteContext, d: typing.Dict[str, typing.Any], guessed_python_types: typing.Dict[str, type] = None
```
Comment (wild-endeavor) [inline flytekit/core/type_engine.py:411]:
```suggestion
        cls, ctx: FlyteContext, d: typing.Dict[str, typing.Any], guessed_python_types: typing.Optional[typing.Dict[str, type]] = None
```
but don't use github to take the suggestion - it doesn't do dco.

File: flytekit/core/type_engine.py [hunk 3]
```diff
@@ -407,17 +407,26 @@ def literal_map_to_kwargs(
...
+    ) -> LiteralMap:
         """
-        Given a dictionary mapping string keys to python values, convert to a LiteralMap.
+        Given a dictionary mapping string keys to python values and a dictionary containing guessed types for such string keys,
+        convert to a LiteralMap.
         """
+        if guessed_python_types is None:
```
Comment (wild-endeavor) [inline flytekit/core/type_engine.py:417]: `guessed_python_types = guessed_python_types or {}`

File: flytekit/remote/remote.py [hunk 4]
```diff
@@ -603,7 +603,8 @@ def _execute(
             disable_all = None
 
         with self.remote_context() as ctx:
-            literal_inputs = TypeEngine.dict_to_literal_map(ctx, inputs)
+            python_types = TypeEngine.guess_python_types(entity.interface.inputs)
```
Comment (wild-endeavor) [inline flytekit/remote/remote.py:606]: this isn't quite what we talked about right?  can we add `guessed_python_interface` to FlyteTask, FlyteLaunchPlan and FlyteWorkflow?  And then in the cases where they are fetched from a local `@task` or `@workflow` the guessed interface is just assigned from the known local python interface.
